ee/institution/admin/saml.sql

```diff
@@ -2,47 +2,84 @@
 SELECT * FROM institutions WHERE id = $id;
 
 -- BLOCK insert_institution_saml_provider
-INSERT INTO saml_providers (
-  institution_id,
-  sso_login_url,
-  issuer,
-  certificate,
-  uid_attribute,
-  uin_attribute,
-  name_attribute,
-  public_key,
-  private_key
-) VALUES (
-  $institution_id,
-  $sso_login_url,
-  $issuer,
-  $certificate,
-  $uid_attribute,
-  $uin_attribute,
-  $name_attribute,
-  COALESCE($public_key, ''),
-  COALESCE($private_key, '')
-) ON CONFLICT (institution_id) DO UPDATE
-SET
-  sso_login_url = EXCLUDED.sso_login_url,
-  issuer = EXCLUDED.issuer,
-  certificate = EXCLUDED.certificate,
-  uid_attribute = EXCLUDED.uid_attribute,
-  uin_attribute = EXCLUDED.uin_attribute,
-  name_attribute = EXCLUDED.name_attribute,
-  public_key = COALESCE($public_key, saml_providers.public_key),
-  private_key = COALESCE($private_key, saml_providers.private_key);
+WITH existing_saml_provider AS (
+  SELECT * FROM saml_providers WHERE institution_id = $institution_id
+), upserted_saml_provider AS (
+  INSERT INTO saml_providers (
+    institution_id,
+    sso_login_url,
+    issuer,
+    certificate,
+    uid_attribute,
+    uin_attribute,
+    name_attribute,
+    public_key,
+    private_key
+  ) VALUES (
+    $institution_id,
+    $sso_login_url,
+    $issuer,
+    $certificate,
+    $uid_attribute,
+    $uin_attribute,
+    $name_attribute,
+    COALESCE($public_key, ''),
+    COALESCE($private_key, '')
+  ) ON CONFLICT (institution_id) DO UPDATE
+  SET
+    sso_login_url = EXCLUDED.sso_login_url,
+    issuer = EXCLUDED.issuer,
+    certificate = EXCLUDED.certificate,
+    uid_attribute = EXCLUDED.uid_attribute,
+    uin_attribute = EXCLUDED.uin_attribute,
+    name_attribute = EXCLUDED.name_attribute,
+    public_key = COALESCE($public_key, saml_providers.public_key),
+    private_key = COALESCE($private_key, saml_providers.private_key)
+  RETURNING *
+), audit_logs_upserted_saml_provider AS (
+  INSERT INTO audit_logs (
+    authn_user_id,
+    table_name,
+    action,
+    institution_id,
+    row_id,
+    old_state,
+    new_state
+  ) SELECT
+    $authn_user_id as authn_user_id,
+    'saml_providers' as table_name,
+    (CASE WHEN esp.id IS NULL THEN 'insert' ELSE 'update' END) as action,
+    $institution_id as institution_id,
+    usp.id AS row_id,
+    (to_jsonb(esp) || jsonb_build_object('private_key', 'REDACTED')) AS old_state,
+    json_build_object(
+      'id', usp.id,
+      'sso_login_url', usp.sso_login_url,
+      'issuer', usp.issuer,
+      'certificate', usp.certificate,
+      'uid_attribute', usp.uid_attribute,
+      'uin_attribute', usp.uin_attribute,
+      'name_attribute', usp.name_attribute,
+      'public_key', usp.public_key,
+      'private_key', 'REDACTED' 
+    ) AS new_state
```

Comment:
Could this be written using a similar expression to the `to_json(esp)` line above?

Reply:
Yes! Let me change that.

Reply:
Changed in eaa756d.